concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/Modules.hs

```diff
@@ -390,24 +390,47 @@ putModulesV0 mods = do
     liftPut $ putWord64be $ LFMB.size mt
     LFMB.mmap_ putModuleV0 mt
 
+-- |Migrate smart contract modules from context @m@ to the context @t m@.
 migrateModules ::
-  (SupportsPersistentModule m, SupportsPersistentModule (t m), SupportMigration m t)
+  forall t m . (SupportsPersistentModule m, SupportsPersistentModule (t m), SupportMigration m t)
   => Modules
   -> t m Modules
 migrateModules mods = do
-    newModulesTable <- LFMB.migrateLFMBTree migrateModule (_modulesTable mods)
+    newModulesTable <- LFMB.migrateLFMBTree migrateCachedModule (_modulesTable mods)
     return
         Modules
             { _modulesMap = _modulesMap mods
             , _modulesTable = newModulesTable
             }
+    where 
+      migrateCachedModule :: CachedModule -> t m CachedModule
+      migrateCachedModule cm = do
+        existingModule <- lift (refLoad cm)
+        case existingModule of
+          ModuleV0 v0 -> migrateModuleV v0
+          ModuleV1 v1 -> migrateModuleV v1
+
+      migrateModuleV :: forall v . (IsWasmVersion v) => ModuleV v -> t m CachedModule
+      migrateModuleV ModuleV{..} = do
```

Comment:
It might be a little simpler to remove `mkModule` and change as follows:
```suggestion
          ModuleV0 v0 -> migrateModuleV ModuleV0 v0
          ModuleV1 v1 -> migrateModuleV ModuleV1 v1

      migrateModuleV :: (ModuleV v -> Module) -> ModuleV v -> t m CachedModule
      migrateModuleV mkModule ModuleV{..} = do
```

Reply:
Not sure. I dislike the need for this embedding as well, but having to provide a callback is to me somewhat less clear than the current situation. So I prefer to keep it as is.